Support for opening a ZIM file by file descriptor

veloman-yunkan opened the pull request:
Fixes #284

Opening a ZIM file via a file descriptor is supported only under Linux. The support is implemented by converting the file descriptor to a path of the form `/dev/fd/<n>`. This approach was taken in order to support the `zim::Item::getDirectAccessInformation()` method, after it [became clear](#issuecomment-739946537) that the only way to deep-copy a file descriptor is to resolve the path of the file underlying it.

Known issues:

1. `zim::FileImpl::verify()` (checksum verification) doesn't work correctly for ZIM files embedded in another file. This can be fixed if the data for checksum computation is read via `zimReader` instead of via an `std::ifstream`. Is that OK?

Pre-clean-up version of this PR was validated in a Kiwix Reader Android app in combination with kiwix/kiwix-lib#429 and kiwix/kiwix-android#2548

kelson42 commented:
@mgautierfr Could you please review this, so we can move on with the features?

kelson42 commented:
@veloman-yunkan Would you be able please to rebase it?

veloman-yunkan commented:
> @veloman-yunkan Would you be able please to rebase it?

Is this going to be merged *before* #454? If not, then I better rebase after the latter is merged.

kelson42 commented:
@veloman-yunkan After discussion with @mgautierfr, we have decided to merge the branch `libzim_next` first. Sorry for the long review time and the rebasing work.

kelson42 commented:
@veloman-yunkan Now it can be rebased.

veloman-yunkan commented:
> @veloman-yunkan Now it can be rebased.

@kelson42 Done

mgautierfr commented on include/zim/archive.h:
```diff
@@ -75,6 +75,16 @@ namespace zim
+       *  @param fd The descriptor of a seekable file with a continuous segment
+       *  representing a complete ZIM archive.
+       */
+      Archive(int fd, offset_type offset, size_type size);
```
I wonder if this is the good api.

Xapian provide a way to open a fd at a specific offset by simply providing `Database(int fd)` constructor. It is to the user of library to seek to the right position before passing the fd.

And even if we provide a `offset` parameter. The fd may still be positioned at a specific offset. We should handle this the proper way (seek it to 0 all the time or taking into account the current position) and document this.

veloman-yunkan replied:
I chose to document the current API

mgautierfr commented on src/fileimpl.cpp:
```diff
@@ -69,16 +69,18 @@ struct NoDelete
+    ASSERT(size, ==, zimFile->fsize());
     return std::make_shared<MultiPartFileReader>(zimFile);
-  else {
+  } else {
```
I missed this in previous commit. But I tend to have braces all the time with `if`/`else`/`while`/... 

It avoid potential future bugs and simplify changes (this one would be simple add of two lines instead of a change of four lines)

mgautierfr commented on test/data/meson.build:
```diff
@@ -1,5 +1,6 @@
 data_files = [
   'small.zim',
+  'small.zim.embedded',
```
(Not specific to this line, but I have to put the comment somewhere)

Why this commit changes all test zim files ?
It should only add a new zim file.

veloman-yunkan replied:
That's an issue with the `create_test_zimfiles` script. It always regenerates all test zimfiles and the new ones don't bitwise match the old ones even if no source data has changed. I usually committed new test zim files selectively. On a couple of occasions (though not in this PR) I also fixed some bugs in the `create_test_zimfiles` script and I decided to re-commit all  ZIM files so that they form a coherent set. I don't remember exactly why I chose to do the same in this commit, but logic suggests that `small.zim.embedded` must be coherent with `small.zim` and the rest (broken ZIM files) were committed to be coherent with the source file from which they were produced.

mgautierfr replied:
> the new ones don't bitwise match the old ones even if no source data has changed

This should not be the case.
There is no reason output changes if input or script don't.

We have more than 1Mb of test zim file.
If we change them every time we add a new test we will pollute our git repository with binary data.

We should maybe move them to git lfs, Or at least only keep only the source zim file and generate other test file at runtime.

mgautierfr commented on src/fileimpl.cpp:
```diff
@@ -63,27 +63,62 @@ sectionSubReader(const FileReader& zimReader, const std::string& sectionName,
+    return std::make_shared<MultiPartFileReader>(zimFile);
+  } else {
+    const auto& firstAndOnlyPart = zimFile->begin()->second;
+    const FileReader::FileHandle fh(&firstAndOnlyPart->fhandle(), NoDelete());
```
Shouldn't we make `FilePart` manage a `std::shared_ptr<FS::FD>` instead of a plain `FS::FD` ?
This would allow us to drop this `NoDelete` shared ptr.

veloman-yunkan replied:
That's definitely a safer solution.

mgautierfr reviewed (changes requested):
A lot of small changes in this commit, I hope I've not missed something.

Few more global comments :
- Now that we have a MultiPartFileReader and FileReader I wonder if we could split `file_reader.(h|cpp)` in two separated files.
- There is this not code question about the api and how to handle fd already positioned. We have to decide something and document it.

veloman-yunkan commented on test/cluster.cpp:
```diff
@@ -320,8 +320,7 @@ TEST(ClusterTest, read_extended_cluster)
 
-  auto filePart = new zim::FilePart<>(fileno(tmpfile));
-  auto fileCompound = std::shared_ptr<zim::FileCompound>(new zim::FileCompound(filePart));
+  auto fileCompound = std::make_shared<zim::FileCompound>(fileno(tmpfile));
```
This comment is not clear

veloman-yunkan commented on test/archive.cpp:
```diff
@@ -325,4 +320,24 @@ ASSERT_EQ(entry1.getIndex(), entry2.getIndex());
+TEST(ZimArchive, openByFD)
+{
+  const zim::Archive archive1("./data/wikibooks_be_all_nopic_2017-02.zim");
+  // XXX: will this work on windows?
```
`FS::openFile()` doesn't provide access to the `int` file descriptor. In this particular case the path is all ascii chars and I don't think that the current directory's path matters (it should be handled inside `_open()`). So I'll leave it as is (more accurately, will squash with e016e36).

mgautierfr replied:
There is the `release` method. Which make the `FD` loose ownership of the handle.
We may add a method to just get the fd.

But I will not block for this. Keeping the `_open` is ok

veloman-yunkan commented:
@mgautierfr I addressed (almost) all review comments as separate commits (also fixed two bugs in `create_test_zimfiles` and `inspectzim` scripts). I will rearrange and squash the commits while rebasing.

mgautierfr reviewed (approved):
I agree with the code but I'm still bothered by this zim file change.

I will not block for this PR but we must find a solution to avoid to put them in the git history.

Can you rebase and squash commits ?

veloman-yunkan commented:
> This should not be the case.
> There is no reason output changes if input or script don't.

One such reason is the changing date that leads to the change of the M/Date metadata item (but it can happen only once every day). Another reason can be the non-determinism in multithreaded processing of input (however I doubt that it takes place for the small input data in our case).

> We have more than 1Mb of test zim file.
> If we change them every time we add a new test we will pollute our git repository with binary data.

While rebasing I added a [change to the create_test_zimfiles](https://github.com/openzim/libzim/pull/449/commits/369fc1ada57b3ff174e45bbfb76d322142533ae2) script that leads to a smaller small.zim (as well as other zim files derived from it).

veloman-yunkan commented:
@mgautierfr Before merging please revisit the known issues listed in the description of the PR (I just added one more known issue).

mgautierfr commented:
> Before merging please revisit the known issues listed in the description of the PR (I just added one more known issue).

You did well to mention that.

The first point is a real issue. On android especially (where the fd constructor will be used), video are read using the `getDirectAccessInformation`.
I discussed a time ago with @macgills and we agree to have `getDirectAccessInformation` return a fd correctly positioned instead of the current pair (and probably rename the method to something more accurate).
I was planning to do this just to have a better API (and a simplified jni wrapper) but I think this is the correct thing to do to have it working with a archive opened on a fd.

> One such reason is the changing date that leads to the change of the M/Date metadata item (but it can happen only once every day). Another reason can be the non-determinism in multithreaded processing of input (however I doubt that it takes place for the small input data in our case).

I missed the important point here : we are recreating the small.zim file with zimwriterfs.
We should not. The test data should not evolved, it is a way to ensure that zimlib has no backward compatibility issue. If we have new feature we want to test, we may generated new test file but we must keep the existing ones (except really specific reasons, dropping zlib support may be one for example, or removing the xapian db).
On top of that, as we generate the zim file with zimwriterfs which use libzim, if there are (coherent) bugs in the reader and writer part, we may not find it.

veloman-yunkan commented:
@mgautierfr So how do we proceed with this PR?

veloman-yunkan commented:
> I discussed a time ago with @macgills and we agree to have `getDirectAccessInformation` return a fd correctly positioned instead of the current pair (and probably rename the method to something more accurate).

I would strongly avoid introducing a non-pure function for that purpose. Modifying the state (position) of a file descriptor in what appears as a query operation may lead to surprises, especially if other code (maybe running in another thread) also depends on the state of the same file-descriptor. libzim readers avoid seeking and instead utiliize `pread()` for a reason. It's safer to return a pair consisting of a file descriptor and an offset value.

mgautierfr commented:
> > I discussed a time ago with @macgills and we agree to have `getDirectAccessInformation` return a fd correctly positioned instead of the current pair (and probably rename the method to something more accurate).
> 
> I would strongly avoid introducing a non-pure function for that purpose. Modifying the state (position) of a file descriptor in what appears as a query operation may lead to surprises, especially if other code (maybe running in another thread) also depends on the state of the same file-descriptor. libzim readers avoid seeking and instead utiliize `pread()` for a reason. It's safer to return a pair consisting of a file descriptor and an offset value.

Hum. I was speaking about a fonction creating/opening a fd and positioning it before returning it.
The returned fd would be totally independent than the internal fd used by libzim.

It would be possible to re-open a file when we have a zim archive opened from a filename.
But I haven't found a way to "fully" duplicate a filedescriptor. (`dup` system function duplicate the fd, but both fd share same internal data (including offset) so it is not what we want).

If we cannot have a really independent fd, yes, we must not position the fd, and we must return a offset. 
But it would be to the user code to create a independent fd (by opening a zim/resource file itself).
We can enforce this when we return a path, but I'm afraid that if we return a fd, we will have user code positioning and reusing our internal fd.

The best API I can think about for now would be to have two methods:
- One returning the offset of the item relative to the start of the zim file (if item is not compressed)
- One returning a filename given a range (offset and size). (if the range is contains in only one file)

For zim archive backed by only one file or a fd, the first method would be enough. The second method would be really usefull to handle split zim file.

(If you have better api idea, you are welcome)

veloman-yunkan commented:
I added 4 draft commits addressing the issue with `getDirectAccessInformation()`.

mgautierfr commented:
I realize this also break xapian database.
We open xapian database by opening a file, positioning the fd and then passing it to xapian.
But here we have no file to open.

I wonder if this idea of opening a zim file by fd is not a wrong idea.
The main purpose is to avoid android app to extract the zim file from a resource file. If we want libzim to read at a known offset in a file, wouldn't be better to simply pass a filename and a offset ?

@macgills, when android give you access to the resource file, does it give you only a fd or the file path to the resource ?

macgills commented:
Opening via fd/inputstream is multifaceted. The android ecosystem is more and more refusing to just tell you that a file is on the filesystem, say in the case that I download a zim from chrome and tap the "open download" button, when I receive this `Intent` in Kiwix app what I get is a "content" URI and I am not aware if this content uri relates to a file on disk.
The other scenario is us embedding a zim file inside kiwix's assets. Unless there have been some changes since I last investigated there is no way of getting a file path. From the app I can open an inputstream or I can acquire an AssetFileDescriptor

mgautierfr commented:
> From the app I can open an inputstream or I can acquire an AssetFileDescriptor

That's a pity.

We may have a solution to reopen a fd but there is no portable way.
On windows, we have to convert the fd to handle, get the filename of the handle and reopen the file.
On linux, we can directly open `/proc/self/fd/<fdnumber>`.
On other unix, it seems we have to look for `/dev/fd/<fdnumber>`. (Also available on linux)

While android is a linux system, they may have done something different.
(@macgills, can you have a look on android what is the content of `proc/<libzimpid>/fd/` ?)

If we go with fd (and it seems necessary for android), we need a to reopen the file to have fulltext search enable.

If we can reopen the file, then we can have a api returning a fd/offset. But here I think that splitting the two features (get the offset of a item and get the file where the range is) in two different methods is better.

macgills commented:
>can you have a look on android what is the content of proc/<libzimpid>/fd/ ?)

Not really get anything obviously libkiwix from `ps`, using the app's pid I get 0-124

kelson42 commented:
@mgautierfr @veloman-yunkan We should in PRIO1 rebase and sort out this ticket/PR. The tickets left https://github.com/openzim/libzim/issues/376 and https://github.com/openzim/zim-tools/issues/69 are PRIO2.

mgautierfr commented:
We need to implement the "reopening of the fd" to have xapian working. (And potentially a getDirectAccessInformation returning a fd/offset).

We can go with getting the filename with `GetFileInformationByHandleEx(FILE_NAME_INFO)` on windows (https://stackoverflow.com/questions/1598586/duplicate-file-descriptor-with-its-own-file-offset) and opening `/dev/fv/<number>` on unix.
But we will have to do some test on android with @macgills to check that everything works correctly.

veloman-yunkan commented:
> We may have a solution to reopen a fd but there is no portable way.
> On windows, we have to convert the fd to handle, get the filename of the handle and reopen the file.
> On linux, we can directly open `/proc/self/fd/<fdnumber>`.
> On other unix, it seems we have to look for `/dev/fd/<fdnumber>`. (Also available on linux)

It looks like the `/dev/fd/<fdnumber>` trick doesn't work under MacOS: https://github.com/openzim/libzim/pull/449/checks?check_run_id=1527077028

kelson42 commented:
@veloman-yunkan Would you be able please to make a Linux only feature? We don't need it for other OSes and I don't expect this to change soon. And we need it urgently on Android Linux.

veloman-yunkan commented:
> We may have a solution to reopen a fd but there is no portable way.
> On windows, we have to convert the fd to handle, get the filename of the handle and reopen the file.
> On linux, we can directly open `/proc/self/fd/<fdnumber>`.
> On other unix, it seems we have to look for `/dev/fd/<fdnumber>`

All this boils down to converting a file-descriptor to a file path. If it works then we can simply drop this PR and instead do that conversion in the `zim::Archive`/`zim::FileImpl` constructor taking a file descriptor.

veloman-yunkan commented:
> > We may have a solution to reopen a fd but there is no portable way.
> > On windows, we have to convert the fd to handle, get the filename of the handle and reopen the file.
> > On linux, we can directly open `/proc/self/fd/<fdnumber>`.
> > On other unix, it seems we have to look for `/dev/fd/<fdnumber>`
> 
> All this boils down to converting a file-descriptor to a file path. If it works then we can simply drop this PR and instead do that conversion in the `zim::Archive`/`zim::FileImpl` constructor taking a file descriptor.

I missed that this PR also adds support for embedded ZIM archives. So we cannot drop it completely.

kelson42 commented:
@veloman-yunkan Thank you, The Windows CI does not pass because of the unit test for fd access (is not activated on linux only?). Surprisingly the macOS test pass.

mgautierfr commented on src/file_part.h:
```diff
@@ -42,10 +42,10 @@ class FilePart {
-        m_filename(""),
-        m_fhandle(std::make_shared<FS::FD>(fd)),
-        m_size(m_fhandle->getSize()) {}
+        FilePart(getFilePathFromFD(fd)) {}
```
We should not resolve the symbolic link.
On linux, it is possible to rename a file or even delete it while it is opened.
The the real path may be wrong when we try to reopen it.
Linux should keep the symbolic link valid in case of rename (I don't know in case of deletion).
We should simply try to open the symlink `/dev/fd/<number>`.

veloman-yunkan replied:
> Linux should keep the symbolic link valid in case of rename (I don't know in case of deletion).

I checked that the symbolic links in /proc/*pid*/fd are indeed updated when a file is renamed. When a file is deleted the symbolic link becomes invalid (with ` (deleted)` appended to that last known filename) but still usable until the process terminates.

veloman-yunkan replied:
Done

veloman-yunkan commented:
I have cleaned up this PR after support for `zim::Item::getDirectAccessInformation()` was implemented in a slightly different way. See updated description.

@mgautierfr I am afraid, you will have to review the PR again in its entirety because of the long pause since last review and a rewrite of its second half.

veloman-yunkan commented:
@mgautierfr @kelson42 I believe that the failure of the ZimArchive unit tests related to search/Xapian functionality is unrelated to this PR as demonstrated by a new [unit test](https://github.com/openzim/libzim/pull/449/commits/db813aaaf856422ae5cba0604fe5e48e0d9f6de5) and reported as issue #471.

veloman-yunkan commented:
The failures of `zim::Search` related unit-tests were indeed caused by a [pre-existing problem in `zim::Search`](https://github.com/openzim/libzim/issues/471#issuecomment-766327169). I provided a [work-around](https://github.com/openzim/libzim/pull/449/commits/441dd2a4016aab32e144f62dfeabae7714e480c5) in the unit-test code, and now we have an [almost clean CI run](https://github.com/openzim/libzim/pull/449/checks?sha=cc80639f6a87e2fdfa90d4a4c5ac3cd6e7f05058), the only failure being [codecov/patch — 80.18% of diff hit (target 90.00%)](https://github.com/openzim/libzim/pull/449/checks?check_run_id=1757705101).

kelson42 commented:
I have rebased this PR on latest git master. @veloman-yunkan @mgautierfr AFAIK this PR is ready to review.

mgautierfr commented on src/item.cpp:
```diff
@@ -91,6 +91,7 @@ std::pair<std::string, offset_type> Item::getDirectAccessInformation() const
-  auto local_offset = full_offset - range.min;
-  return std::make_pair(part->filename(), offset_type(local_offset));
+  const offset_type local_offset(full_offset - range.min);
+  const offset_type offset = local_offset + m_file->getArchiveStartOffset().v;
```
Shouldn't we add the `archiveStartOffset` to the `full_offset` before we search for the part ?

We have a `archiveStartOffset` different of zero only when we open a embedded archive (so only one part) for now but we may have in the future some kind of opening of embedded archive in a split file.

veloman-yunkan replied:
Makes sense

mgautierfr commented on include/zim/archive.h:
```diff
@@ -67,6 +67,7 @@ namespace zim
        */
       explicit Archive(const std::string& fname);
 
+#ifndef _WIN32
```
Please add a comment in the docstring.
We use the linux build to generate the doc. We must tell user that the feature is not enable on windows.

veloman-yunkan replied:
Done

mgautierfr reviewed (changes requested):
Two small changes (I think) and we are good.

kelson42 commented:
@veloman-yunkan Regarding the codecov/patch failing test, is the 90% limit too high or should we have a better coverage?

veloman-yunkan commented:
> @veloman-yunkan Regarding the codecov/patch failing test, is the 90% limit too high or should we have a better coverage?

The 90% limit is not too high. In this particular case the difficulty of meeting this requirement is because there are fat blocks of error handling code which is not covered (see https://codecov.io/gh/openzim/libzim/pull/449/src/src/file_reader.cpp?before=src/file_reader.cpp). In order to cover those lines we need to simulate an IO failure.

mgautierfr reviewed (approved):
We are good.
Please rebase-fixup and we can merge.

kelson42 commented:
@mgautierfr This is already properly rebased. Merging...

veloman-yunkan commented:
> @mgautierfr This is already properly rebased. Merging...

This needed a rebase in order to squash the fixup commits...

kelson42 commented:
@veloman-yunkan Yes, my bad. We will have to leave with it.
